app/api/contests/set_additional_time.go

```diff
@@ -57,59 +61,132 @@ import (
+	if err != nil {
+		return 0, 0, 0, service.ErrInvalidRequest(err)
+	}
+	groupID, err = service.ResolveURLQueryPathInt64Field(r, "group_id")
+	if err != nil {
+		return 0, 0, 0, service.ErrInvalidRequest(err)
+	}
+	seconds, err = service.ResolveURLQueryGetInt64Field(r, "seconds")
+	if err != nil {
+		return 0, 0, 0, service.ErrInvalidRequest(err)
+	}
+	const maxSeconds = 838*3600 + 59*60 + 59
+	// 838:59:59 is the maximum possible TIME value in MySQL
+	if seconds < -maxSeconds || maxSeconds < seconds {
+		return 0, 0, 0, service.ErrInvalidRequest(fmt.Errorf("'seconds' should be between %d and %d", -maxSeconds, maxSeconds))
+	}
+	return itemID, groupID, seconds, service.NoError
+}
+
+func setAdditionalTimeForGroupInContest(
+	store *database.DataStore, groupID, itemID, participantsGroupID, durationInSeconds, additionalTimeInSeconds int64) {
+	groupContestItemStore := store.GroupContestItems()
+	scope := groupContestItemStore.Where("group_id = ?", groupID).Where("item_id = ?", itemID)
+	found, err := scope.WithWriteLock().HasRows()
+	service.MustNotBeError(err)
+	if found {
+		service.MustNotBeError(scope.UpdateColumn("additional_time",
+			gorm.Expr("SEC_TO_TIME(?)", additionalTimeInSeconds)).Error())
+	} else if additionalTimeInSeconds != 0 {
+		service.MustNotBeError(groupContestItemStore.Exec(
+			"INSERT INTO groups_contest_items (group_id, item_id, additional_time) VALUES(?, ?, SEC_TO_TIME(?))",
+			groupID, itemID, additionalTimeInSeconds).Error())
+	}
+
+	service.MustNotBeError(store.Exec("DROP TEMPORARY TABLE IF EXISTS total_additional_times").Error())
+	service.MustNotBeError(store.Exec(`
+		CREATE TEMPORARY TABLE total_additional_times (
+			PRIMARY KEY child_group_id (child_group_id)
+		)
+		?`,
+		store.GroupGroups().
+			Where("groups_groups.type"+database.GroupRelationIsActiveCondition).
+			Where("groups_groups.parent_group_id = ?", participantsGroupID).
+			Joins(`
+				JOIN groups_ancestors AS changed_group_descendants
+					ON changed_group_descendants.child_group_id = groups_groups.child_group_id AND
+						changed_group_descendants.ancestor_group_id = ?`, groupID).
+			Joins(`
+				JOIN contest_participations
+					ON contest_participations.group_id = groups_groups.child_group_id AND
+						contest_participations.entered_at IS NOT NULL AND
+						contest_participations.item_id = ?`, itemID).
+			Joins("JOIN groups_ancestors ON groups_ancestors.child_group_id = groups_groups.child_group_id").
+			Joins(`
+				JOIN groups_contest_items
+					ON groups_contest_items.group_id = groups_ancestors.ancestor_group_id AND
+						groups_contest_items.item_id = contest_participations.item_id`).
+			Group("groups_groups.child_group_id").
+			Select(`
+				groups_groups.child_group_id,
+				IFNULL(SUM(TIME_TO_SEC(groups_contest_items.additional_time)), 0) AS total_additional_time`).
+			WithWriteLock().QueryExpr()).Error())
+
+	result := store.Exec(`
+		UPDATE groups_groups
```

Comment:
If I do the UPDATE using the SELECT request directly inside it, it seems to works on my db. (so no needing the temp table) I just replaced `JOIN total_additional_times` by `JOIN (SELECT ... ) AS total_additional_times`
Could you check? 
If so, probably there are also ways to write it in a bit cleaner way.

Reply:
Do you have triggers in your DB? (Initially I implemented this with nested select, but that didn't work)

Reply:
("Can't update table in stored function/trigger because it is already used by statement which invoked this stored function/trigger")

Reply:
Yes, I have the triggers and applied the migrations, strange. 🤔 The tables are empty, but this error should not depend on the actual data.

Reply:
Just tried this again: ```Data: {"success":false,"message":"Internal Server Error","error_text":"Error 1442: Can't update table 'groups_ancestors' in stored function/trigger because it is already used by statement which invoked this stored function/trigger."}```